cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,489 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#define __STDC_FORMAT_MACROS
+
+#include "cvmfs_config.h"
+#include "sqlitemem.h"
+
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <new>
+
+#include "smalloc.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace sqlite {
+
+uint32_t MemoryManager::MallocArena::GetSize(void *ptr) const {
+  assert(Contains(ptr));
+
+  ReservedBlockCtl *block_ctl = reinterpret_cast<ReservedBlockCtl *>(
+    reinterpret_cast<char *>(ptr) - sizeof(ReservedBlockCtl));
+  int32_t size = -(block_ctl->size);
+  assert(size > 1);
+  return size - sizeof(ReservedBlockCtl) - 1;
+}
+
+/**
+ * Creates a free block at the place of the reserved block ptr points into.
+ * The free block might need to be merged with adjacent lower and/or upper
+ * blocks.  In these cases, the corresponding blocks are removed from the list
+ * of available blocks.  Every allocated block has a predecessor and a
+ * successor in the arena.  The newly created free block is added to the end of
+ * the list of available blocks.
+ */
+void MemoryManager::MallocArena::Free(void *ptr) {
+  assert(Contains(ptr));
+
+  no_reserved_--;
+
+  ReservedBlockCtl *block_ctl = reinterpret_cast<ReservedBlockCtl *>(
+    reinterpret_cast<char *>(ptr) - sizeof(ReservedBlockCtl));
+  char prior_tag = *(reinterpret_cast<char *>(block_ctl) - 1);
```

Comment:
This is getting the tag from the previous block, isn't it? What happens if we are dealing with the first block in the arena?